File: PythonAlgorithmInterView/문자열조작/팰린드롬.js
/**
 * @param {string} s
 * @return {boolean}
 * 125번 문제.
 */
 var isPalindrome = function(s) {
    let replace_str_list = s.toLowerCase().replace(/[^a-z0-9]/g,'').split('');
    
    while(replace_str_list.length > 1){
        if(replace_str_list.pop() != replace_str_list.shift()){
            return false;
        }
    }

    return true;
    
};

console.log(isPalindrome("A man, a plan, a canal: Panama"));
console.log(isPalindrome("race a car"));
console.log(isPalindrome("0P"));

File: PythonAlgorithmInterView/문자열조작/팰린드롬.py
import re
import collections


# 락의 해결방안.
def rock_solution(str) -> bool:
    reg_str_list = re.findall('[a-zA-Z]', str);

    str1 = ''.join(reg_str_list).upper()
    reg_str_list.reverse()
    str2 = ''.join(reg_str_list).upper()

    return str1 == str2


# 상길북에서의 해결방안.
def real_solution(s) -> bool:
    strs = []

    for char in s:
        if char.isalnum():
            strs.append(char.lower())

    while len(strs) > 1:
        if strs.pop(0) != strs.pop():
            return False

    return True

# optimization 된 기능
def optimization_soultion( s: str) -> bool:
    strs = collections.deque()

    for char in s :
        if char.isalnum():
            strs.append(char.lower())
    while len(strs) > 1:
        if strs.popleft() != strs.pop():
            return False
    return True

if __name__ == '__main__':
    print(optimization_soultion("A man, a plan, a canal: Panama"))
    # print(real_solution("race a car"))
